File: epi_judge_java/epi/SearchEntryEqualToIndex.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;
import epi.test_framework.TimedExecutor;

import java.util.List;

public class SearchEntryEqualToIndex {

    public static int searchEntryEqualToItsIndex(List<Integer> list) {
        int left = 0;
        int right = list.size() - 1;

        while (left <= right) {
            int midIndex = left + (right - left) / 2;
            if (list.get(midIndex) == midIndex) {
                return midIndex;
            } else if (list.get(midIndex) < midIndex) {
                left = midIndex + 1;
            } else {
                right = midIndex - 1;
            }
        }

        return -1;
    }

    @EpiTest(testDataFile = "search_entry_equal_to_index.tsv")
    public static void searchEntryEqualToItsIndexWrapper(TimedExecutor executor,
                                                         List<Integer> A)
            throws Exception {
        int result = executor.run(() -> searchEntryEqualToItsIndex(A));

        if (result != -1) {
            if (A.get(result) != result) {
                throw new TestFailure("Entry does not equal to its index");
            }
        } else {
            for (int i = 0; i < A.size(); ++i) {
                if (A.get(i) == i) {
                    throw new TestFailure("There are entries which equal to its index");
                }
            }
        }
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "SearchEntryEqualToIndex.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/search_entry_equal_to_index.cc
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using std::vector;

int SearchEntryEqualToItsIndex(const vector<int> &nums) {
    int left = -1;
    int right = nums.size();
    while (left + 1 < right) {
        int candidate = (left + right) / 2;
        if (nums[candidate] == candidate) return candidate;
        if (nums[candidate] < candidate) left = candidate;
        else right = candidate;
    }
    return -1;
}

void SearchEntryEqualToItsIndexWrapper(TimedExecutor &executor,
                                       const vector<int> &A) {
    int result = executor.Run([&] { return SearchEntryEqualToItsIndex(A); });

    if (result != -1) {
        if (A[result] != result) {
            throw TestFailure("Entry does not equal to its index");
        }
    } else {
        for (int i = 0; i < A.size(); ++i) {
            if (A[i] == i) {
                throw TestFailure("There are entries which equal to its index");
            }
        }
    }
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"executor", "A"};
    return GenericTestMain(
        args, "search_entry_equal_to_index.cc", "search_entry_equal_to_index.tsv",
        &SearchEntryEqualToItsIndexWrapper, DefaultComparator{}, param_names);
}
